servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java

```diff
@@ -186,6 +198,65 @@ public final HttpServerBuilder appendConnectionAcceptorFilter(final ConnectionAc
+     * {@link #listenStreaming(StreamingHttpService)} before it is used by the server.
+     * <p>
+     * The order of execution of these filters are in order of append, before the filters added with
+     * {@link #appendServiceFilter(StreamingHttpServiceFilterFactory)}. If 3 filters are added as follows:
+     * <pre>
+     *     builder.appendServiceFilter(filter1).appendNonOffloadingServiceFilter(filter2).appendServiceFilter(filter3)
+     * </pre>
+     * accepting a request by a service wrapped by this filter chain, the order of invocation of these filters will be:
+     * <pre>
+     *     filter2 ⇒ filter1 ⇒ filter3 ⇒ service
+     * </pre>
+     *
+     * @param factory {@link StreamingHttpServiceFilterFactory} to append.
+     * @return {@code this}
+     * @throws IllegalArgumentException if the provided filter requires offloading.
+     */
+    public final HttpServerBuilder appendNonOffloadingServiceFilter(final StreamingHttpServiceFilterFactory factory) {
+        noOffloadServiceFilters.add(checkNonOffloading("Non-offloading filter",
+                HttpExecutionStrategies.defaultStrategy(), factory));
+        return this;
+    }
+
+    /**
+     * Appends a non-offloading filter to the chain of filters used to decorate the {@link StreamingHttpService} used
+     * by this builder, for every request that passes the provided {@link Predicate}. Filters added via this method
+     * will be executed before offloading occurs and before filters appended via
+     * {@link #appendServiceFilter(StreamingHttpServiceFilterFactory)}.
+     * <p>
+     * Note this method will be used to decorate the {@link StreamingHttpService} passed to
+     * {@link #listenStreaming(StreamingHttpService)} before it is used by the server.
+     * The order of execution of these filters are in order of append, before the filters added with
+     * {@link #appendServiceFilter(StreamingHttpServiceFilterFactory)}. If 3 filters are added as follows:
+     * <pre>
+     *     builder.appendServiceFilter(filter1).appendNonOffloadingServiceFilter(filter2).appendServiceFilter(filter3)
+     * </pre>
+     * accepting a request by a service wrapped by this filter chain, the order of invocation of these filters will be:
+     * <pre>
+     *     filter2 ⇒ filter1 ⇒ filter3 ⇒ service
+     * </pre>
+     *
+     * @param predicate the {@link Predicate} to test if the filter must be applied. This must not block.
+     * @param factory {@link StreamingHttpServiceFilterFactory} to append.
+     * @return {@code this}
+     * @throws IllegalArgumentException if the provided filter or predicate requires offloading.
+     */
+    public final HttpServerBuilder appendNonOffloadingServiceFilter(final Predicate<StreamingHttpRequest> predicate,
+                                                                    final StreamingHttpServiceFilterFactory factory) {
+        checkNonOffloading("Non-offloading predicate", noOffloadsStrategy(), predicate);
```

Comment:
This is a great catch that we also have to check the predicate! However, we don't do anything for a predicate in `appendServiceFilter`. Any plans to improve that?

Reply:
Good comment. This comment suggested a bunch of changes to `ConditionalFilterFactory` and related classes, now in PR #1730.